File: min-failing-path.py
# Time Complexity: O(n*n) where A has n cols and n rows
# Space Complexity: O(1) - using input array as dp array
# Did it run on Leetcode: Yes
# Did you face any issues: No
# Approach:
#     - Decrease and conquer the problem
#     - Define an objective function f(i,j) as the minimum cost at any index i,j
#     - This depends on the values the previous row that differ by +/-1
#     - f(i,j) = min(f(i-1, j), f(i-1, j-1), f(i-1,j))

class Solution(object):
    def minFallingPathSum(self, A):
        """
        :type A: List[List[int]]
        :rtype: int
        """
        # Using input array as dp array
        for i in range(1, len(A)):
            for j in range(len(A[0])):
                if j == 0:  # Left boundary, consider only j and j+1
                    candidate = min(A[i-1][j], A[i-1][j+1])
                elif j == len(A[0])-1:  # Right boundry
                    candidate = min(A[i-1][j-1], A[i-1][j])
                else:
                    candidate = min(A[i-1][j-1], A[i-1][j], A[i-1][j+1])
                A[i][j] += candidate

        return min(A[-1])

File: MinFailingPath.cpp
// Time Complexity : (row x col)
// Space Complexity : (row x col)
// Did this code successfully run on Leetcode : yes
// Any problem you faced while coding this : NO


class Solution {
public:
    vector<vector<int>> dir = {{1,-1},{1,0},{1,1}};
    int helper(vector<vector<int>>& matrix, int r, int c, vector<vector<int>> &dp) {
        if(r == matrix.size()-1) {
            return matrix[r][c];
        }
        if(dp[r][c] != -1) {
            return dp[r][c];
        }
        int ans = INT_MAX;
        for(auto d : dir) {
            int nr = r + d[0];
            int nc = c + d[1];
            if(nr >= 0 && nc >= 0 && nr < matrix.size() && nc < matrix[0].size()) {
                ans = min(ans, helper(matrix, nr, nc, dp));
            }
        }
        return dp[r][c] = matrix[r][c] + ans;
    }
    int minFallingPathSum(vector<vector<int>>& matrix) {
        vector<vector<int>> dp(matrix.size(), vector<int>(matrix[0].size(), -1));
        int ans = INT_MAX;
        for(int i=0;i<matrix[0].size();i++) {
            ans = min(ans, helper(matrix, 0, i, dp));
        }
        return ans;
    }
};